---
title: "analysis NMD"
output: html_document
date: "2023-08-02"
---

## Load required libraries
```{r}
library(markdown)
library(rmarkdown)
library(NMF)
library(ggplot2)
library(readxl)
library(tidyverse)
library(DESeq2)
library(pheatmap)
```

## Data loading and processing
The following CircRNACount file contains multiple samples some of which might be filtered
```{r eval=FALSE}
data_count <- read.table("CircRNACount", sep = "\t", header = T)
```

Read the CircCoordinates file which also stores the annotation information for circularRNAs. Create id of circRNA that includes annotation column because there are multiple entries with same coordinates but annotated differently and also have different counts. So keep these entries separate
```{r eval=FALSE}
data_coordinate <- read.table("CircCoordinates", sep = "\t", header = T)
```

Join the two dataframes
```{r eval=FALSE}
data_join <- cbind(data_count, data_coordinate)
data_join$id <- paste(data_join$Chr, data_join$Start, data_join$End, data_join$Gene , sep = "_")
```


Read the sample ID excel file
```{r eval=FALSE}
samples <- read_xlsx(path = "data.xlsx", sheet = "Fractionation Seq (LS)", col_names = T)
```

Grep on samples from fraction-seq data from https://redmine.dieterichlab.org/projects/circularrnaejcnmd/wiki
```{r eval=FALSE}
selected_samples <- unlist(lapply(samples$CCG_Sample_ID, function(x) grep(x, names(data_join), value = T)))
data_count_fseq <- data_join[,c("id", selected_samples)]
rownames(data_count_fseq) <- data_count_fseq$id
data_count_fseq$id <- NULL
```

Remove "_STARmappingChimeric.out.junction" from column names aka sample names
```{r eval=FALSE}
colnames(data_count_fseq) = gsub("_STARmappingChimeric.out.junction", "", colnames(data_count_fseq))
```

Add counts from two lanes to create a final count matrix
```{r eval=FALSE}
test <- lapply(samples$CCG_Sample_ID, function(x) as.data.frame(rowSums(data_count_fseq[,grepl(x,colnames(data_count_fseq))])))
dim(bind_cols(test))
final_data_count_fseq <- bind_cols(test)
colnames(final_data_count_fseq) <- samples$CCG_Sample_ID
```

## Statistics
First perform the PCA analysis based on circular RNAs count matrix
```{r eval=FALSE}
nsubnum <- sum( rowMeans( counts(dds, normalized=TRUE)) > 5 )
vsd <- vst(dds, blind=FALSE, nsub = nsubnum)
rld <- rlog(dds, blind=FALSE)
head(assay(vsd), 3)
```

PCA plot
```{r fig.align='center', fig.height=8, fig.width=8}
pcadata <- plotPCA(vsd, intgroup = c("Knockdown", "Replicate"), returnData = T)
percentVar <- round(100 * attr(pcadata, "percentVar"))
ggplot(pcadata, aes(PC1, PC2, color=Knockdown, shape=Replicate)) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme(aspect.ratio=1)
```

Sample-sample distance matrix
```{r eval=FALSE}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$Condition
colnames(sampleDistMatrix) <- NULL
```

Plot the heatmap of count matrix
```{r fig.align='center', fig.height=10, fig.width=11}
print(pheatmap(sampleDistMatrix, clustering_distance_rows=sampleDists, clustering_distance_cols=sampleDists)) #,col=colors))
```

There is clear separation between nuclear and cytoplasmic samples and also between different knockdown conditions. Next, perform the differential expression analysis on this count matrix to find differentially expressed transcript between various knockdown experiments. In this case, perform the nuclear versis cytoplasmic differential expression for all conditions i.e. knockdown experiments one-by-one.
```{r eval=FALSE}
coldata <- samples[,c("CCG_Sample_ID", "Condition", "Knockdown", "Replicate")]
rownames(coldata) <- coldata$CCG_Sample_ID
list_res <- list()
for (each_cond in unique(coldata$Knockdown)){
  print(each_cond)
  sub_samples <- coldata$CCG_Sample_ID[coldata$Knockdown == each_cond]
  dds <- DESeqDataSetFromMatrix(final_data_count_fseq[,sub_samples]+1, colData = coldata[sub_samples,], design = ~ Replicate)
  dds <- estimateSizeFactors(dds)
  dds <- DESeq(dds)
  res <- results(dds)
  list_res[[each_cond]] <- as.data.frame(res)
}
```

