---
title: "Codes for re-analysis of differential expression of circular and linear RNAs using Circtest"
output: html_document
date: "2024-01-11"
---

## Loading required libraries

```{r}
library(CircTest)
library(ggplot2)
library(dplyr)
library(rmarkdown)
library(markdown)
library(readxl)
library(DESeq2)
library(gprofiler2)
```

# Analysis
## Functions

Function to prepare the data in format to be inputted to CircTest
```{r eval=FALSE}
prepare_data <- function(data_file, data_file_coord, controls, samples){
  data <- read.table(data_file, header = T)
  data_coord <- read.table(data_file_coord, header = T, sep = "\t")
  if (nrow(data) != nrow(data_coord)){
    print("Number of lines do not match with Coordinate file")
    stop()
  }
  data$id <- paste(data$Chr, data$Start, data$End, data_coord$Gene, sep = "_")
  data$id <- make.unique(data$id, sep = "_")
  
  ind_control <- unlist(lapply(controls, function(x) grep(x, colnames(data))))
  ind_sample <- unlist(lapply(samples, function(x) grep(x, colnames(data))))
  data <- data[,c(ncol(data), ind_control, ind_sample)]
  colnames(data) <- c("ID", "Control_1", "Control_2", "Control_3", "Sample_1", "Sample_2", "Sample_3")
  return(data)
}
```

Function for CircTest calling
```{r eval=FALSE}
call_circtest <- function(circrnacount, linearcount){
  # # Add pseudocounts to count dataframes
  ind <- which(sapply(circrnacount, is.integer))
  circrnacount[ind] <- circrnacount[ind] + 1
  linearcount[ind] <- linearcount[ind] + 1
  # filtering counts
  Circ_filtered <- Circ.filter(circ = circrnacount, linear = linearcount, Nreplicates = 3, filter.sample = 3, filter.count = 5, percentage = 0.01, circle_description = 1)
  Linear_filtered <- linearcount[rownames(Circ_filtered),]
  # circtest command
  circtest_result <- Circ.test(Circ_filtered, Linear_filtered, group=c(rep(1,3),rep(2,3)), circle_description = 1, alpha = 1)
}
```

Function for DESeq2 differential expression
```{r eval=FALSE}
call_deseq2 <- function(data_count){
  rownames(data_count) <- data_count$ID
  data_count$ID <- NULL
  coldata <- as.data.frame(colnames(data_count))
  colnames(coldata) <- c("Sample")
  coldata$Type <- unlist(lapply(coldata$Sample, function(x) strsplit(x, "_")[[1]][1]))
  coldata$Replicate <- unlist(lapply(coldata$Sample, function(x) strsplit(x, "_")[[1]][2]))
  rownames(coldata) <- coldata$Sample
  
  # DE analysis
  dds <- DESeqDataSetFromMatrix(countData = data_count+1, colData = coldata, design = ~ Type)
  dds <- DESeq(dds)
  res <- as.data.frame(results(dds))
  res$padj[is.na(res$padj)] <- 1 
  signif_res <- res[res$padj < 0.05,]
  
  # #PCA plot
  # rld <- rlog(dds, blind=TRUE)
  # plotPCA(rld, intgroup="Type") + coord_fixed()
  return(res)
}
```

```{r}
de_patterns <- function(circles_de, linear_de){
  #circles_de <- deseq2_circle
  #linear_de <- deseq2_linear
  df <- matrix(ncol=3, nrow=nrow(circles_de))
  for (i in 1:nrow(circles_de)){
    if (! rownames(circles_de)[i] %in% rownames(linear_de)){
      #print("Unique DE circle")
      df[i,] <- c(rownames(circles_de)[i], circles_de$log2FoldChange[i], 0)
    } else {
      #print("This one")
      df[i,] <- c(rownames(circles_de)[i], circles_de$log2FoldChange[i], linear_de$log2FoldChange[rownames(linear_de) %in% rownames(circles_de)[i]])
    }
  }
  df <- as.data.frame(df)
  colnames(df) <- c("ID", "Log2FC_Circle", "Log2FC_Linear")
  df$Log2FC_Circle <- as.double(df$Log2FC_Circle)
  df$Log2FC_Linear <- as.double(df$Log2FC_Linear)
  df$flag_circle <- ifelse(df$Log2FC_Circle > 0, "UP", "DOWN")
  df$flag_linear <- ifelse(df$Log2FC_Linear > 0, "UP", "DOWN")
  df$flag_linear <- ifelse(df$Log2FC_Linear == 0, "NULL", df$flag_linear)
  df$flag_combined <- paste(df$flag_circle, df$flag_linear, sep = "_")
  print(table(df$flag_combined))
  return(df)
}
```

## Circtest
Read the file with metadata for circtests
```{r eval=FALSE}
metadata <- read_xlsx("circtest_sample_metadata.xlsx", sheet = "Sheet1")
```

Calling function for each sample and getting the dataframe
```{r eval=FALSE}
list_data_circle <- list()   # list that stores all data objects that will be used for circtest and deseq2
list_data_linear <- list()
list_circtest <- list()
for (i in 1:nrow(metadata)){
  # forming file, control and sample names
  circcount_file <- metadata$CircRNACount[i]
  linearcount_file <- metadata$LinearRNACount[i]
  coordinate_file <- metadata$CircCoordinate[i]
  control_names <- strsplit(metadata$Control[i], ";")[[1]]
  sample_names <- strsplit(metadata$Sample[i], ";")[[1]]
  print(paste(i, metadata$Data[i], metadata$Experiment[i])) #, circcount_file, linearcount_file))
  # calling function for preparing the data
  circrnacount <- prepare_data(circcount_file, coordinate_file, control_names, sample_names)
  linearcount <- prepare_data(linearcount_file, coordinate_file, control_names, sample_names)
  list_data_circle[[paste(metadata$Data[i], metadata$Experiment[i], sep = "__")]] <- circrnacount
  list_data_linear[[paste(metadata$Data[i], metadata$Experiment[i], sep = "__")]] <- linearcount
  # calling function for circtest 
  res_circtest <- call_circtest(circrnacount, linearcount)
  list_circtest[[paste(metadata$Data[i], metadata$Experiment[i], sep = "__")]] <- res_circtest
}
```

Printing number of circles passed the criteria and significant circles
```{r}
for (each_id in names(list_circtest)){
  print(paste(each_id, nrow(list_circtest[[each_id]]$summary_table), nrow(list_circtest[[each_id]]$summary_table[list_circtest[[each_id]]$summary_table$sig_p < 0.05,])))
}
```

## DESeq2
### Circular RNAs differential expression
```{r eval=FALSE}
list_deseq2_circle <- list()
for (each_id in names(list_data_circle)){
  res_deseq2 <- call_deseq2(list_data_circle[[each_id]])
  list_deseq2_circle[[each_id]] <- res_deseq2
  signif_res_deseq2 <- res_deseq2[res_deseq2$padj < 0.05,]
  print(paste(each_id, nrow(signif_res_deseq2)))
}
```

### Linear RNAs differential expression
```{r eval=FALSE}
list_deseq2_linear <- list()
for (each_id in names(list_data_linear)){
  res_deseq2 <- call_deseq2(list_data_linear[[each_id]])
  list_deseq2_linear[[each_id]] <- res_deseq2
  signif_res_deseq2 <- res_deseq2[res_deseq2$padj < 0.05,]
  print(paste(each_id, nrow(signif_res_deseq2)))
}
```

## Circtest versus DESeq2 comparison
Circtest finds out circular RNAs who's expression changes along with their linear host genes. So compare the differentially expressing circles only in DESeq2 who's linear RNAs are not DE.
```{r}
#`%nin%` = Negate(`%in%`)
for (each_id in names(list_circtest)){
  print(each_id)
  deseq2_circle <- list_deseq2_circle[[each_id]][list_deseq2_circle[[each_id]]$padj < 0.05,]
  deseq2_linear <- list_deseq2_linear[[each_id]][list_deseq2_linear[[each_id]]$padj < 0.05,]
  if (nrow(deseq2_circle) == 0){
    next
  }
  circtest_both <- list_circtest[[each_id]]$summary_table[list_circtest[[each_id]]$summary_table$sig_p < 0.05,]
  # for every circle in deseq2, see if it is present in linear deseq2 list and see the log2FC enrichment
  df_comparison <- de_patterns(deseq2_circle, deseq2_linear)
}
```

Also, create the significant results list separately for later
```{r eval=FALSE}
find_sig <- function(data){
  data <- data[data$padj < 0.05,]
  return(data)
}
list_deseq2_circle_sig <- lapply(list_deseq2_circle, find_sig)
list_deseq2_linear_sig <- lapply(list_deseq2_linear, find_sig)
```


Combined circtest results list into one dataframe. Also add DESeq2 results as separate columns
```{r}
list_circtest_combined <- list()
for (each_id in names(list_circtest)){
  print(each_id)
  each_list <- list_circtest[[each_id]]
  data <- cbind(each_list$summary_table[,c("ID", "sig_p")], each_list$ratios)
  rownames(data) <- data$ID
  colnames(data) <- c("ID", "pvalue", "group_1_ratio_mean", "group_2_ratio_mean")
  data$deseq2_circles <- ifelse(data$ID %in% rownames(list_deseq2_circle_sig[[each_id]]), TRUE, FALSE)
  data$deseq2_linear <- ifelse(data$ID %in% rownames(list_deseq2_linear_sig[[each_id]]), TRUE, FALSE)
  #print(table(data$deseq2_circles))
  # unique circles from deseq2 not in circtest
  rownames(list_deseq2_circle_sig[[each_id]]) %in% each_list$summary_table$ID
  list_circtest_combined[[each_id]] <- data
  
  # correlation between deseq2 circle and linear fold changes
  print(dim(list_deseq2_circle_sig[[each_id]]))
  list_deseq2_circle_sig[[each_id]]$log2FoldChange_linear <- list_deseq2_linear_sig[[each_id]]$log2FoldChange[match(rownames(list_deseq2_circle_sig[[each_id]]), rownames(list_deseq2_linear_sig[[each_id]]))]
  list_deseq2_circle_sig[[each_id]]$log2FoldChange_linear[is.na(list_deseq2_circle_sig[[each_id]]$log2FoldChange_linear)] <- 0
  
  print(cor.test(list_deseq2_circle_sig[[each_id]]$log2FoldChange_linear, list_deseq2_circle_sig[[each_id]]$log2FoldChange))
}
```


## Genome-wide differential expression analysis
Function for DESeq2 analysis for all genes
```{r}
metadata_all <- read_xlsx("circtest_sample_metadata.xlsx", sheet = "DESeq2_all")
```

Creating count matrix
```{r}
for (i in 1:nrow(metadata_all)){
  print(metadata_all$ID[i])
  sample_ids <- c(strsplit(metadata_all$Control[i], ";")[[1]], strsplit(metadata_all$Sample[i], ";")[[1]])
  count <- 1
  for (each_id in sample_ids){
    print(each_id)
    mapping_file <- grep(paste(each_id, "_STARmapping", sep = ""), list.files(paste(metadata_all$Path[i], "mapping/", sep = ""), pattern = "ReadsPerGene.out.tab", full.names = T, recursive = T), value = T)
    mapping_out <- read.table(mapping_file)
    mapping_out <- mapping_out[grepl("N_", mapping_out$V1) ==FALSE,]  # remove unwanted stats line
    rownames(mapping_out) <- mapping_out$V1
    #mapping_out$V1 <- NULL
    if (count == 1){
      data_count <- mapping_out[,c("V1", "V2")]
      data_count$V1 <- NULL
    } else {
      temp <- data.frame(mapping_out[,c("V2")])
      rownames(temp) <- rownames(mapping_out)
      data_count <- merge(data_count, temp, by=0, all=TRUE)
      rownames(data_count) <- data_count$Row.names
      data_count$Row.names <- NULL
    }
    count <- count + 1 
  }
  colnames(data_count) <- c("Control_1", "Control_2", "Control_3", "Sample_1", "Sample_2", "Sample_3")
}

metadata <- as.data.frame(colnames(data_count))
colnames(metadata) <- c("Sample")
metadata$Type <- unlist(lapply(metadata$Sample, function(x) strsplit(x, "_")[[1]][1]))
metadata$Replicate <- unlist(lapply(metadata$Sample, function(x) strsplit(x, "_")[[1]][2]))
rownames(metadata) <- metadata$Sample

dds <- DESeqDataSetFromMatrix(countData = data_count+1, colData = metadata, design = ~ Type)

```




